ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/SnapServer.java

```diff
@@ -14,66 +14,321 @@
+    // TODO: drop to TRACE
+    LOGGER.atInfo()
+        .setMessage("Receive get bytecodes message for {} hashes")
+        .addArgument(codeHashes.hashes()::size)
+        .log();
+
+    // there is no worldstate root or block header for us to use, so default to head.  This
+    // can cause problems for self-destructed contracts pre-shanghai.  for now since this impl
+    // is deferring to #5889, we can just get any flat code storage and know we are not deleting
+    // code for now.
+    return worldStateStorageProvider
+        .apply(Optional.empty())
+        .map(
+            storage -> {
+              List<Bytes> codeBytes = new ArrayDeque<>();
+              for (Bytes32 codeHash : codeHashes.hashes()) {
+                storage.getCode(codeHash, null).ifPresent(codeBytes::add);
+              }
+              return ByteCodesMessage.create(codeBytes);
+            })
+        .orElse(EMPTY_BYTE_CODES_MESSAGE);
   }
 
-  private MessageData constructGetBytecodesResponse(
-      final WorldStateArchive worldStateArchive, final MessageData message) {
-    // TODO implement
-    return ByteCodesMessage.create(new ArrayDeque<>());
+  MessageData constructGetTrieNodesResponse(final MessageData message) {
+    final GetTrieNodesMessage getTrieNodesMessage = GetTrieNodesMessage.readFrom(message);
+    final GetTrieNodesMessage.TrieNodesPaths triePaths = getTrieNodesMessage.paths(true);
+    // TODO: drop to TRACE
+    LOGGER
+        .atInfo()
+        .setMessage("Receive get trie nodes message of size {}")
+        .addArgument(() -> triePaths.paths().size())
+        .log();
+
+    // TODO: implement limits
+    return worldStateStorageProvider
+        .apply(Optional.of(triePaths.worldStateRootHash()))
+        .map(
+            storage -> {
+              ArrayList<Bytes> trieNodes = new ArrayList<>();
+              for (var triePath : triePaths.paths()) {
+                // first element in paths is account
+                if (triePath.size() == 1) {
+                  // if there is only one path, presume it should be compact encoded account path
+                  storage.getTrieNodeUnsafe(triePath.get(0)).ifPresent(trieNodes::add);
+                } else {
+                  // otherwise the first element should be account hash, and subsequent paths
+                  // are compact encoded account storage paths
+
+                  final Bytes accountPrefix = triePath.get(0);
+
+                  triePath.subList(1, triePath.size()).stream()
+                      .forEach(
+                          path ->
+                              storage
+                                  .getTrieNodeUnsafe(Bytes.concatenate(accountPrefix, path))
+                                  .ifPresent(trieNodes::add));
+                }
+              }
+              return TrieNodesMessage.create(trieNodes);
+            })
+        .orElse(EMPTY_TRIE_NODES_MESSAGE);
+  }
+
```

Comment:
```suggestion
  MessageData constructGetTrieNodesResponse(final MessageData message) {
    final GetTrieNodesMessage getTrieNodesMessage = GetTrieNodesMessage.readFrom(message);
    final GetTrieNodesMessage.TrieNodesPaths triePaths = getTrieNodesMessage.paths(true);
    // TODO: drop to TRACE
    LOGGER
        .atInfo()
        .setMessage("Receive get trie nodes message of size {}")
        .addArgument(() -> triePaths.paths().size())
        .log();

    // TODO: implement limits
    return worldStateStorageProvider
        .apply(Optional.of(triePaths.worldStateRootHash()))
        .map(
            storage -> {
              ArrayList<Bytes> trieNodes = new ArrayList<>();
              for (var triePath : triePaths.paths()) {
                // first element in paths is account
                if (triePath.size() == 1) {
                  // if there is only one path, presume it should be compact encoded account path
                  storage.getTrieNodeUnsafe(CompactEncoding.decode(triePath.get(0))).ifPresent(trieNodes::add);
                } else {
                  // otherwise the first element should be account hash, and subsequent paths
                  // are compact encoded account storage paths

                  final Bytes accountPrefix = triePath.get(0);

                  triePath.subList(1, triePath.size()).stream()
                      .forEach(
                          path ->
                              storage
                                  .getTrieNodeUnsafe(Bytes.concatenate(accountPrefix, CompactEncoding.decode(path)))
                                  .ifPresent(trieNodes::add));
                }
              }
              return TrieNodesMessage.create(trieNodes);
            })
        .orElse(EMPTY_TRIE_NODES_MESSAGE);
  }

```